src/listener.js

```diff
@@ -1,183 +1,263 @@
-    if (this.connected) { return }
-    this.address = address
-    this.aid = String(this.address.decapsulate('p2p-stardust'))
+  /**
+   * Read discovery messages
+   * @private
+   */
+  async _discoverPeers () {
+    if (!this.isConnected) {
+      return
+    }
 
-    log('connecting to %s', address)
+    const baseAddr = this.address.decapsulate('p2p-stardust')
+    let message
 
-    let conn = await this.client.switch.dial(address.decapsulate('p2p-stardust'))
-    const muxed = await this.client.switch.wrapInMuxer(conn, false)
+    try {
+      message = await this.wrappedStream.readPB(Discovery)
+
+      // Proof still connected
+      this.wrappedStream.writePB({}, DiscoveryAck)
+    } catch (err) {
+      // listener was already closed
+      if (!this.isConnected) {
+        return
+      }
 
-    conn = await prom(cb => muxed.once('stream', s => cb(null, s)))
-    conn = await this.client.switch.negotiateProtocol(conn, '/p2p/stardust/0.1.0')
-    const rpc = LP(conn)
+      log('failed to read discovery: %s', err.stack)
+      log('assume disconnected!')
 
-    log('performing challenge')
+      this.isConnected = false
+      this.serverConnection = null
+      const wrappedStream = this.wrappedStream.unwrap()
+      wrappedStream.sink([])
 
-    const random = crypto.randomBytes(128)
-    rpc.writeProto(JoinInit, { random128: random, peerID: this.client.id.toJSON() })
+      log('reconnecting')
 
-    log('sent rand')
+      try {
+        await this.listen(this.address)
+        log('reconnected!')
+      } catch (e) {
+        log('reconnect failed: %s', e.stack)
+      }
 
-    const { error, saltEncrypted } = await rpc.readProto(JoinChallenge)
-    if (error) { translateAndThrow(error) }
-    const saltSecret = await prom(cb => this.client.id.privKey.decrypt(saltEncrypted, cb))
+      return
+    }
 
-    const solution = sha5(random, saltSecret)
-    rpc.writeProto(JoinChallengeSolution, { solution })
+    if (!this.client.discovery._isStarted) {
+      log('reading discovery, but tossing data since it\'s not enabled')
+      return
+    }
 
-    const { error: error2 } = await rpc.readProto(JoinVerify)
-    if (error2) { translateAndThrow(error2) }
+    log('reading discovery')
+    for (const id of message.ids) {
+      try {
+        const pi = new PeerInfo(new PeerId(id))
 
-    log('connected')
+        if (pi.id.toB58String() !== this.client.id.toB58String()) {
+          pi.multiaddrs.add(baseAddr.encapsulate('/p2p-stardust/p2p/' + pi.id.toB58String()))
 
-    this.connected = true
-    this.muxed = muxed
-    this.rpc = rpc
-    this.client.connections[this.aid] = this
+          this.client.discovery.emit('peer', pi)
+        }
+      } catch (err) {
+        log.error('invalid peer discovered', err)
+      }
+    }
 
-    muxed.on('stream', this.handler)
-    this._readDiscovery()
+    this._timeoutId = setTimeout(() => this._discoverPeers(), this.options.discoveryInterval || 10000) // cooldown
```

Comment:
This will queue up messages if the server's discovery interval is higher. We can either exchange the interval via the join success message or just directly read (though the second one is better since connection lags can get this out of hand and break the ACK mechanism - thus disconnecting the peer)

Reply:
Yes, I agree that we can remove the timeout and just block on the stream read